subsys/bluetooth/controller/hci/hci.c

```diff
@@ -4798,6 +4800,151 @@ static void vs_read_tx_power_level(struct net_buf *buf, struct net_buf **evt)
+ */
+#if defined(CONFIG_CPU_CORTEX_M)
+typedef struct bt_hci_vs_fata_error_cpu_data_cortex_m bt_hci_vs_fatal_error_cpu_data;
+
+static void vs_err_fatal_cpu_data_fill(bt_hci_vs_fatal_error_cpu_data *cpu_data,
+				       const z_arch_esf_t *esf)
+{
+	cpu_data->a1 = sys_cpu_to_le32(esf->basic.a1);
+	cpu_data->a2 = sys_cpu_to_le32(esf->basic.a2);
+	cpu_data->a3 = sys_cpu_to_le32(esf->basic.a3);
+	cpu_data->a4 = sys_cpu_to_le32(esf->basic.a4);
+	cpu_data->ip = sys_cpu_to_le32(esf->basic.ip);
+	cpu_data->lr = sys_cpu_to_le32(esf->basic.lr);
+	cpu_data->xpsr = sys_cpu_to_le32(esf->basic.xpsr);
+}
+#endif /* CONFIG_CPU_CORTEX_M */
+
+static struct net_buf *vs_err_evt_create(uint8_t subevt, uint8_t len)
+{
+	struct net_buf *buf;
+
+	buf = net_buf_alloc(&vs_err_tx_pool, K_FOREVER);
+	if (buf) {
+		struct bt_hci_evt_le_meta_event *me;
+		struct bt_hci_evt_hdr *hdr;
+
+		net_buf_reserve(buf, BT_BUF_RESERVE);
+		bt_buf_set_type(buf, BT_BUF_EVT);
+
+		hdr = net_buf_add(buf, sizeof(*hdr));
+		hdr->evt = BT_HCI_EVT_VENDOR;
+		hdr->len = len + sizeof(*me);
+
+		me = net_buf_add(buf, sizeof(*me));
+		me->subevent = subevt;
+	}
+
+	return buf;
+}
+
+struct net_buf *hci_vs_err_stack_frame(unsigned int reason, const z_arch_esf_t *esf)
+{
+	/* Prepare vendor specific HCI Fatal Error event */
+	struct bt_hci_vs_fatal_error_stack_frame *sf;
+	bt_hci_vs_fatal_error_cpu_data *cpu_data;
+	struct net_buf *buf;
+
+	buf = vs_err_evt_create(BT_HCI_EVT_VS_ERROR_DATA_TYPE_STACK_FRAME,
+				sizeof(*sf) + sizeof(*cpu_data));
+	if(buf != NULL) {
+		sf = net_buf_add(buf, (sizeof(*sf) + sizeof(*cpu_data)));
+		sf->reason = sys_cpu_to_le32(reason);
+		sf->cpu_type = BT_HCI_EVT_VS_ERROR_CPU_TYPE_CORTEX_M;
+
+		vs_err_fatal_cpu_data_fill(
+			(bt_hci_vs_fatal_error_cpu_data *)sf->cpu_data, esf);
+	} else {
+		BT_ERR("Can't create HCI Fatal Error event");
+	}
+
+	return buf;
+}
+
+static struct net_buf *hci_vs_err_trace_create(uint8_t data_type,
+					       const char *file_path,
+					       uint32_t line, uint64_t pc)
+{
+	uint32_t len = 0U, pos = 0U;
+	struct net_buf *buf = NULL;
+
+	if (file_path) {
+		/* Extract file name from a path */
+		while (file_path[len] != '\0') {
+			if (file_path[len] == '/') {
+				pos = len + 1;
+			}
+			len++;
+		}
+		file_path += pos;
+		len -= pos;
+
+		/* If file name was found in file_path, in other words: file_path is not empty
+		 * string and is not `foo/bar/`.
+		 */
+		if (len) {
+			/* Total data length: len = file name strlen + \0 + sizeof(line number)
+			 * Maximum length of an HCI event data is BT_BUF_RX_SIZE. If total data
+			 * length exceeds this maximum, truncate file name.
+			 */
+			uint32_t data_len = 1 + sizeof(line);
+
+			/* If a buffer is created for a TRACE data, include sizeof(pc) in total
+			 * length.
+			 */
+			if (data_type == BT_HCI_EVT_VS_ERROR_DATA_TYPE_TRACE) {
+				data_len += sizeof(pc);
+			}
+
+			if (data_len + len > BT_BUF_RX_SIZE) {
+				uint32_t overflow_len = len + data_len - BT_BUF_RX_SIZE;
```

Comment:
Shouldn't it actually be `BT_BUF_EVT_RX_SIZE`?

Reply:
No, `BT_BUF_EVT_RX_SIZE` is used in buffers memory definition:
```
NET_BUF_POOL_FIXED_DEFINE(vs_err_tx_pool, 1, BT_BUF_RX_SIZE, 8, NULL);
```
`BT_BUF_EVT_RX_SIZE` may equal `BT_BUF_RX_SIZE` if `BT_BUF_EVT_RX_SIZE` is the biggest one of all other kinds of buffers used in HCI.

Reply:
Right, but if `BT_BUF_EVT_RX_SIZE` is e.g. 150, and `BT_BUF_ACL_RX_SIZE` is 200, then `BT_BUF_RX_SIZE` is 200, but how can we then create an event here of size (`BT_BUF_RX_SIZE`/200) if the host uses `BT_BUF_EVT_RX_SIZE`/150 for the event pool?

```c
NET_BUF_POOL_FIXED_DEFINE(evt_pool, CONFIG_BT_BUF_EVT_RX_COUNT,
			  BT_BUF_EVT_RX_SIZE, 8,
			  NULL);
```

Wouldn't events > 150 get discarded then?

Reply:
Well Host cannot use an event buffer smaller than 255 bytes if it is not 100% that all events fit into that size. This is an optimization for cooperation with well known Controller. 
https://github.com/zephyrproject-rtos/zephyr/blob/0cb22dfa1de5be2b828d67a9748b157105e3ba43/subsys/bluetooth/host/hci_core.c#L119-L125
and:
https://github.com/zephyrproject-rtos/zephyr/blob/d96f361175ba759d7344edbea3a3b5f1e766cba8/include/zephyr/bluetooth/buf.h#L90-L92
Nevertheless for optimization purpose I've changed the `evt_pool` definition to use `BT_BUF_EVT_RX_SIZE` because it is used only for events. That can save couple of bytes in case the event buffer size is not maximum allowed.